# US01 - 2 
## Python
### Data Bank 

The goal is to feel an xlsx with the GDP of a list of countries. The list of contries are on a datasheet. It's possible to get the gdp on internet thanks a API  (application programming interface). 


**1. List of the countries**

We need to get the list of countries from the datasheet. We import the data of the datasheet in a dataframe

In [3]:
#import a library to open xlsx
import pandas as pd 

In [4]:
#read_excel read the xlsx. It contains many parameters. 
#see more: https://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_excel.html
df = pd.read_excel('Data-sheet-STUDENTS.xlsx',
             header=2,
             usecols=[1],
             names=['countries'] 
             )

In [5]:
df['countries']

0                            Afghanistan
1                                Albania
2                                Armenia
3                             Azerbaijan
4                                Bahrain
5                             Bangladesh
6                                Belarus
7                                 Bhutan
8                 Bosnia and Herzegovina
9                                 Brunei
10                              Bulgaria
11                              Cambodia
12                                 China
13                               Croatia
14                        Czech Republic
15                            East Timor
16                                 Egypt
17                               Estonia
18                      France (Not BRI)
19                               Georgia
20                     Germany (not BRI)
21                               Hungary
22                                 India
23                             Indonesia
24              

**2. Get the ISO code**

We need know to get the GDP thanks to a API: the API world bank accept the ISO code and not the properly name of the country. So we need to change the name of country into the ISO code. For example: France = fr

To do so we can use this API : https://restcountries.eu/rest/v2/name/france?format=json and a dictionnary which contains all the countries from our datasheet

In [6]:
countries={}
for country in df['countries'] :
    countries[country] = {}
countries

{'Afghanistan': {},
 'Albania': {},
 'Armenia': {},
 'Azerbaijan': {},
 'Bahrain': {},
 'Bangladesh': {},
 'Belarus': {},
 'Bhutan': {},
 'Bosnia and Herzegovina': {},
 'Brunei': {},
 'Bulgaria': {},
 'Cambodia': {},
 'China': {},
 'Croatia': {},
 'Czech Republic': {},
 'East Timor': {},
 'Egypt': {},
 'Estonia': {},
 'France (Not BRI)': {},
 'Georgia': {},
 'Germany (not BRI)': {},
 'Hungary': {},
 'India': {},
 'Indonesia': {},
 'Iran': {},
 'Iraq': {},
 'Israel': {},
 'Jordan': {},
 'Kazakhstan': {},
 'Kuwait': {},
 'Kyrgyzstan': {},
 'Laos': {},
 'Latvia': {},
 'Lebanon': {},
 'Lithuania': {},
 'Macedonia': {},
 'Malaysia': {},
 'Maldives': {},
 'Moldova': {},
 'Mongolia': {},
 'Montenegro': {},
 'Myanmar': {},
 'Nepal': {},
 'Oman': {},
 'Pakistan': {},
 'Philippines': {},
 'Poland': {},
 'Qatar': {},
 'Romania': {},
 'Russia': {},
 'Saudi Arabia': {},
 'Serbia': {},
 'Singapore': {},
 'Slovakia': {},
 'Slovenia': {},
 'Sri Lanka': {},
 'Syria': {},
 'Tajikistan': {},
 'Thailiand'

In [7]:
import requests
from time import sleep
#from tqdm import tnrange, tqdm_notebook

#from i in tnrange(len(countries.keys())):
for country in countries.keys():
    try: 
        #counrty = list(countries.keys())[i]
        if countries[country] == {}: 
            url = ' https://restcountries.eu/rest/v2/name/'+country
            r = requests.get(url)
            data = r.json();
            countries[country]['alpha2Code'] = data[0]['alpha2Code']
        
    except KeyError: 
        print(data, country)
        sleep(5)


Afghanistan AF
Albania AL
Armenia AM
Azerbaijan AZ
Bahrain BH
Bangladesh BD
Belarus BY
Bhutan BT
Bosnia and Herzegovina BA
Brunei BN
Bulgaria BG
Cambodia KH
China CN
Croatia HR
Czech Republic CZ
East Timor TL
Egypt EG
Estonia EE
{'status': 404, 'message': 'Not Found'}
Georgia GE
{'status': 404, 'message': 'Not Found'}
Hungary HU
India IO
Indonesia ID
Iran IR
Iraq IQ
Israel IL
Jordan JO
Kazakhstan KZ
Kuwait KW
Kyrgyzstan KG
Laos LA
Latvia LV
Lebanon LB
Lithuania LT
Macedonia MK
Malaysia MY
Maldives MV
Moldova MD
Mongolia MN
Montenegro ME
Myanmar MM
Nepal NP
Oman OM
Pakistan PK
Philippines PH
Poland PL
Qatar QA
Romania RO
Russia RU
Saudi Arabia SA
Serbia RS
Singapore SG
Slovakia SK
Slovenia SI
Sri Lanka LK
Syria SY
Tajikistan TJ
{'status': 404, 'message': 'Not Found'}
{'status': 404, 'message': 'Not Found'}
Turkey TR
Turkmenistan TM
{'status': 404, 'message': 'Not Found'}
{'status': 404, 'message': 'Not Found'}
Ukraine UA
Uzbekistan UZ
Vietnam VN
Yemen YE


We can see that there is a problem for 6 country : {'status': 404, 'message': 'Not Found'}. We need to dig more futher to find why it doesn't work, find a solution and the ISO code.

In [8]:
for country, code in countries.items():
    countryCorrect = ''
    if code == {}:
        if 'not bri' in country.lower(): #france, germany, united kingdom, united states of america
            countryCorrect  = country.split('(')[0].strip()
        elif country == 'Thailiand': #thailand
            countryCorrect = 'Thailand'
        else: #the united arab emirates
            countryCorrect = country[4:]
        try: 
        #counrty = list(countries.keys())[i]
            if countries[country] == {}: 
                url = ' https://restcountries.eu/rest/v2/name/'+countryCorrect
                r = requests.get(url)
                data = r.json();
                countries[country]['alpha2Code'] = data[0]['alpha2Code']
        
        except KeyError: 
            print(data, country)
            sleep(5)    

            

So now, know, we have all our countries name with their ISO code in our dictionnary of countries:

In [9]:
countries

{'Afghanistan': {'alpha2Code': 'AF'},
 'Albania': {'alpha2Code': 'AL'},
 'Armenia': {'alpha2Code': 'AM'},
 'Azerbaijan': {'alpha2Code': 'AZ'},
 'Bahrain': {'alpha2Code': 'BH'},
 'Bangladesh': {'alpha2Code': 'BD'},
 'Belarus': {'alpha2Code': 'BY'},
 'Bhutan': {'alpha2Code': 'BT'},
 'Bosnia and Herzegovina': {'alpha2Code': 'BA'},
 'Brunei': {'alpha2Code': 'BN'},
 'Bulgaria': {'alpha2Code': 'BG'},
 'Cambodia': {'alpha2Code': 'KH'},
 'China': {'alpha2Code': 'CN'},
 'Croatia': {'alpha2Code': 'HR'},
 'Czech Republic': {'alpha2Code': 'CZ'},
 'East Timor': {'alpha2Code': 'TL'},
 'Egypt': {'alpha2Code': 'EG'},
 'Estonia': {'alpha2Code': 'EE'},
 'France (Not BRI)': {'alpha2Code': 'FR'},
 'Georgia': {'alpha2Code': 'GE'},
 'Germany (not BRI)': {'alpha2Code': 'DE'},
 'Hungary': {'alpha2Code': 'HU'},
 'India': {'alpha2Code': 'IO'},
 'Indonesia': {'alpha2Code': 'ID'},
 'Iran': {'alpha2Code': 'IR'},
 'Iraq': {'alpha2Code': 'IQ'},
 'Israel': {'alpha2Code': 'IL'},
 'Jordan': {'alpha2Code': 'JO'},
 'Kaza

We can see that the ISO code for India isn't right (IO), so we change it manually. 

In [10]:
countries['India']['alpha2Code'] = 'IN'

**3. GDP for each country**

We need to use another API to get the GDP from world bank thanks to the ISO code. 

In [12]:
for k,v in countries.items():
    try: 
        url = 'http://api.worldbank.org/v2/countries/'+v['alpha2Code']+'/indicators/NY.GDP.MKTP.CD?format=json'
        r = requests.get(url)
        data = r.json()
        countries[k]['gdp'] = float(data[1][0]['value'])
        sleep(2)
    except TypeError: 
        print('typeerror: ', k)        
    except IndexError: 
        print('indexerror: ', k)    
        
countries        

typeerror:  Syria
typeerror:  Yemen


{'Afghanistan': {'alpha2Code': 'AF', 'gdp': 20815300220.0428},
 'Albania': {'alpha2Code': 'AL', 'gdp': 13039352743.9616},
 'Armenia': {'alpha2Code': 'AM', 'gdp': 11536590635.8268},
 'Azerbaijan': {'alpha2Code': 'AZ', 'gdp': 40747792237.9735},
 'Bahrain': {'alpha2Code': 'BH', 'gdp': 35307127659.5745},
 'Bangladesh': {'alpha2Code': 'BD', 'gdp': 249723887765.42},
 'Belarus': {'alpha2Code': 'BY', 'gdp': 54442374372.5094},
 'Bhutan': {'alpha2Code': 'BT', 'gdp': 2511852941.17647},
 'Bosnia and Herzegovina': {'alpha2Code': 'BA', 'gdp': 18168579578.1952},
 'Brunei': {'alpha2Code': 'BN', 'gdp': 12128089001.5858},
 'Bulgaria': {'alpha2Code': 'BG', 'gdp': 56831518294.4397},
 'Cambodia': {'alpha2Code': 'KH', 'gdp': 22158209502.6391},
 'China': {'alpha2Code': 'CN', 'gdp': 12237700479375.0},
 'Croatia': {'alpha2Code': 'HR', 'gdp': 54849180228.8716},
 'Czech Republic': {'alpha2Code': 'CZ', 'gdp': 215725534372.371},
 'East Timor': {'alpha2Code': 'TL', 'gdp': 2954620999.58441},
 'Egypt': {'alpha2Code':

We can see that Yemen and Syria don't have a GPD (because of the war). So we can add a GDP at "None"

In [15]:
countries['Yemen']['gdp'] = 'None'
countries['Syria']['gdp'] = 'None'

**4. Add the gdp value into the datasheet**

First we add the GDP in the dataframe. Secondly, we have to add into the datasheet

In [16]:
#add into the dataframe
s = pd.Series([v['gdp'] for k,v in countries.items()])
df['gdp'] = s
df

countries          gdp
0                          Afghanistan  2.08153e+10
1                              Albania  1.30394e+10
2                              Armenia  1.15366e+10
3                           Azerbaijan  4.07478e+10
4                              Bahrain  3.53071e+10
5                           Bangladesh  2.49724e+11
6                              Belarus  5.44424e+10
7                               Bhutan  2.51185e+09
8               Bosnia and Herzegovina  1.81686e+10
9                               Brunei  1.21281e+10
10                            Bulgaria  5.68315e+10
11                            Cambodia  2.21582e+10
12                               China  1.22377e+13
13                             Croatia  5.48492e+10
14                      Czech Republic  2.15726e+11
15                          East Timor  2.95462e+09
16                               Egypt  2.35369e+11
17                             Estonia  2.59211e+10
18                    France (Not BRI)   2.5825e+12
19                             Georgia  1.51593e+10
20                   Germany (not BRI)  3.67744e+12
21                             Hungary  1.39135e+11
22                               India  2.59749e+12
23                           Indonesia  1.01554e+12
24                                Iran  4.39514e+11
25                                Iraq  1.97716e+11
26                              Israel  3.50851e+11
27                              Jordan  4.00683e+10
28                          Kazakhstan  1.59407e+11
29                              Kuwait  1.20126e+11
..                                 ...          ...
38                             Moldova  8.12849e+09
39                            Mongolia   1.1488e+10
40                          Montenegro  4.77409e+09
41                             Myanmar  6.93221e+10
42                               Nepal   2.4472e+10
43                                Oman  7.26427e+10
44                            Pakistan  3.04952e+11
45                         Philippines  3.13595e+11
46                              Poland   5.2451e+11
47                               Qatar  1.67605e+11
48                             Romania  2.11803e+11
49                              Russia  1.57752e+12
50                        Saudi Arabia  6.83827e+11
51                              Serbia  4.14316e+10
52                           Singapore  3.23907e+11
53                            Slovakia   9.5769e+10
54                            Slovenia  4.87697e+10
55                           Sri Lanka  8.71747e+10
56                               Syria         None
57                          Tajikistan   7.1457e+09
58                           Thailiand  4.55221e+11
59            The United Arab Emirates  3.82575e+11
60                              Turkey  8.51102e+11
61                        Turkmenistan  4.23554e+10
62            United Kingdom (Not BRI)  2.62243e+12
63  United States of America (Not BRI)  1.93906e+13
64                             Ukraine  1.12154e+11
65                          Uzbekistan  4.87177e+10
66                             Vietnam  2.23864e+11
67                               Yemen         None

[68 rows x 2 columns]

In [19]:
#add a graph from the dataframe
%matplotlib inline
import matplotlib.pyplot as plt

df.plot()

TypeError: Empty 'DataFrame': no numeric data to plot

In [18]:
#add into the datasheet
writer = pd.ExcelWriter('output.xlsx')
df.to_excel(writer,'Sheet1')
writer.save()